Change Fatal to Error

liangchenye opened the pull request:
Fix `dies on the first error` issue mentioned in https://github.com/opencontainers/ocitools/issues/98.
Check/validate all the tests, output all the error messages.

It works better with https://github.com/opencontainers/ocitools/pull/97.

zenlinTechnofreak commented:
Can U wirte a readme about how to use generate to different cmds?

wangkirin commented:
@zenlinTechnofreak OK，I‘m finishing the last cmds, but meet some problem, working on it

wking committed:
generate: Drop runtime-supplied devices and mounts

Since opencontainers/specs#164 (part of the v0.1.1 release [1]), Linux
runtimes are required to supply these devices and mounts.  Specifying
them explicitly in the config should have no effect, and just makes
the config longer than it needs to be.

[1]: https://github.com/opencontainers/specs/blob/v0.1.1/config-linux.md#default-devices-and-file-systems

Signed-off-by: W. Trevor King <wking@tremily.us>

liangchenye committed:
add contrib directory and mv get-stage3 to that

Signed-off-by: liangchenye <liangchenye@huawei.com>

liangchenye committed:
rename to rootfs-builder

Signed-off-by: liangchenye <liangchenye@huawei.com>

Commit:
Merge pull request #17 from liangchenye/contrib

add contrib directory and mv get-stage3 to that

liangchenye committed:
add bundle validate as sub-command

Signed-off-by: liangchenye <liangchenye@huawei.com>

liangchenye committed:
bundle validate update to 0.3.0

Signed-off-by: liangchenye <liangchenye@huawei.com>

liangchenye committed:
update generate.go to 0.3.0

Signed-off-by: liangchenye <liangchenye@huawei.com>

liangchenye committed:
update Godeps of opencontainers/specs to 0.3.0

Signed-off-by: liangchenye <liangchenye@huawei.com>

liangchenye committed:
update cmd/runtimetest to 0.3.0

Signed-off-by: liangchenye <liangchenye@huawei.com>

Commit:
Merge pull request #20 from liangchenye/bundle

add bundle validate as sub-command

Commit:
Merge pull request #2 from wking/drop-runtime-supplied-devices-and-mounts

generate: Drop devices and mounts

liangchenye committed:
update to 0.4.0; modify codes accordingly

Signed-off-by: liangchenye <liangchenye@huawei.com>

liangchenye committed:
Create MAINTAINERS

Since this project is based on ocitools and OCT, 
I propose to add Mrunal and me as the initial maintainers. 

Signed-off-by: liangchenye liangchenye@huawei.

Commit:
Merge pull request #23 from liangchenye/maintainers

Add Maintainer

Commit:
Merge pull request #21 from liangchenye/master

update Godeps to specs-go version; modify codes accordingly

mrunalp committed:
Fixes to ocitools generate to make it work with runc again

This commit adds back the mounts since runc doesn't default these yet.

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

mrunalp committed:
Pass uuid to runc start per spec to make runtimes tests work again

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

Commit:
Merge pull request #24 from mrunalp/fix_generate

Fixes to ocitools generate to make it work with runc again

Commit:
Merge pull request #25 from mrunalp/fix_runtime_test

Pass uuid to runc start per spec to make runtime tests work again

mrunalp committed:
Add support for the no new privs flag

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

mrunalp committed:
Fix format flag (reported by go vet)

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

Commit:
Merge pull request #26 from mrunalp/no_new_priv

Add support for the no new privs flag

Commit:
Merge pull request #27 from mrunalp/fix_format

Fix format flag (reported by go vet)

liangchenye committed:
use opencontainer/runtime-spec

Signed-off-by: liang chenye <liangchenye@huawei.com>

mrunalp committed:
Add support for sysctls to generate

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

liangchenye committed:
use rspec; change 'Check' to 'check' as go vet warns

Signed-off-by: liang chenye <liangchenye@huawei.com>

liangchenye committed:
remove go get vet

Signed-off-by: liang chenye <liangchenye@huawei.com>

Commit:
Merge pull request #28 from liangchenye/master

use opencontainer/runtime-spec

Commit:
Merge pull request #29 from mrunalp/sysctl

Add support for sysctls to generate

rhatdan committed:
Add man pages for ocitools

Also fix up Makefile so that it will install ocitools and man pages.

Signed-off-by: Dan Walsh <dwalsh@redhat.com>

rhatdan committed:
Rename bvalidate to validate

Signed-off-by: Dan Walsh <dwalsh@redhat.com>

Commit:
Merge pull request #31 from rhatdan/validate

Rename bvalidate to validate

Commit:
Merge pull request #33 from rhatdan/man

Add man pages for ocitools

rhatdan committed:
Add bash completions

Signed-off-by: Dan Walsh <dwalsh@redhat.com>

Commit:
Merge pull request #32 from rhatdan/completions

Add bash completions

mrunalp committed:
Add support for enabling/disabling tty to generate

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

rhatdan committed:
Fixes for bash completions

Signed-off-by: Dan Walsh <dwalsh@redhat.com>

mrunalp committed:
Add support for using a base template

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

Commit:
Merge pull request #34 from mrunalp/generate_tty

Add support for enabling/disabling tty to generate

Commit:
Merge pull request #35 from mrunalp/template

Add support for using a base template

mrunalp committed:
Don't set hostname by default

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

mrunalp committed:
Only set boolean parameters if they are passed as flags

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

Commit:
Merge pull request #36 from mrunalp/default_hostname

Don't set hostname by default

Commit:
Merge pull request #37 from mrunalp/fix_bool

Only set boolean parameters if they are passed as flags

mrunalp committed:
Add support for annotations(labels) to generate

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

Commit:
Merge pull request #38 from mrunalp/labels

Add support for annotations(labels) to generate

mrunalp committed:
Update runtime-spec to 6e08c6983ef8c2173f10ca09266907d4e9e71716

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

mrunalp committed:
Add support for mount label to generate

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

Commit:
Merge pull request #39 from mrunalp/mount_label

Add support for mount label to generate

mrunalp committed:
Add convenience flag seccomp-allow to generate

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

wking committed:
man/ocitools-generate: Use [OPTIONS] in the synopsis

There are so many of these, I think it's time to follow grep(1) and
friends and punt the individual details to the OPTIONS section.

Signed-off-by: W. Trevor King <wking@tremily.us>

wking committed:
CONTRIBUTING: Document Signed-off-by and the DCO

This project seems to have been operating under the assumption that
the usual Signed-off-by <-> DCO mapping applied.  But it's nice to
make that explicit.  This commit copies the current DCO section from
runtime-spec, but puts it in CONTRIBUTING.md for a more native GitHub
experience [1].

Looking through the history of the repo, the only non-merge commits
without sign-offs are by Mrunal:

* 5286bb06 (Update runtime validation section, 2016-01-19)
* b71f0452 (Update README.md, 2015-11-09)
* 2c0e9e20 (Update help for generate, 2015-11-05)
* 850abe9f (Update README.md, 2015-09-15)
* bb903899 (Initial commit, 2015-09-15)

who clearly doesn't have a problem signing for the rest of his work
and is unlikely to have stolen those contributions from somewhere else
;).

[1]: https://github.com/blog/1184-contributing-guidelines

Signed-off-by: W. Trevor King <wking@tremily.us>

wking committed:
man/ocitools-generate: Alphabetize the OPTIONS section

With --uid so far from --gid, it seems clear that alphabetical order
was the original goal.  Make some adjustments to bring the current
content back in line.

Signed-off-by: W. Trevor King <wking@tremily.us>

wking committed:
man/ocitools-generate: Document --template

Which landed in a937f7a4 (Add support for using a base template,
2016-04-13, #35) and still needs docs [1].

[1]: https://github.com/opencontainers/ocitools/pull/35#issuecomment-210191802

Signed-off-by: W. Trevor King <wking@tremily.us>

wking committed:
man/ocitools-generate: Fix three-space indents for some shell commands

These were typos from 78e69bf5 (Add man pages for ocitools,
2016-04-09, #33).

Signed-off-by: W. Trevor King <wking@tremily.us>

wking committed:
README: Punt detailed command description to the man pages

This makes it easier to get a project overview from the README, and
avoids the tedium of keeping parallel sets of detailed docs up to
date.

Signed-off-by: W. Trevor King <wking@tremily.us>

wking committed:
man/ocitools-generate.1: Copy-edits for the description

Also adjust the copy of this text in the README.

I've left the reference-style links out of the man page, since
go-md2man doesn't seem to be able to handle them, and instead creates
troff with just the link targets:

  \[la]man/ocitools-generate.1.md\[ra] generates a
  \[la]https://github.com/opencontainers/runtime-spec/blob/master/config.md\[ra] for an
  ...

Signed-off-by: W. Trevor King <wking@tremily.us>

wking committed:
generate: Change --mount-cgroups default to 'no'

Since a937f7a4 (Add support for using a base template, 2016-04-13,
--mount-cgroups is not 'no', that will lead to additional cgroup
entries in the mount array (even if there was already a cgroup entry
in the mount array).  Instead of writing logic to find and clobber
existing cgroup entries, just make the default the no-op setting.

Signed-off-by: W. Trevor King <wking@tremily.us>

Commit:
validate: add Hooks validation

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

Commit:
Merge pull request #43 from wking/document-dco

CONTRIBUTING: Document Signed-off-by and the DCO

Commit:
Merge pull request #48 from wking/mount-cgroups-default-to-no

generate: Change --mount-cgroups default to 'no'

Commit:
Merge pull request #46 from wking/man-generate-shell-indents

man/ocitools-generate: Fix three-space indents for some shell commands

Commit:
Merge pull request #47 from wking/readme-link-man-pages

README: Punt detailed command description to the man pages

Commit:
Merge pull request #40 from mrunalp/seccomp_allow

Add convenience flag seccomp-allow to generate

Commit:
Merge pull request #44 from wking/man-generate-alphabetize-options

man/ocitools-generate: Alphabetize the OPTIONS section

RobDolinMS committed:
[ ReadMe ] Spelling fix

"OCI compatable runtime" --> "OCI compatible runtime"

Signed-off-by: Rob Dolin <RobDolin@microsoft.com>

Commit:
Merge pull request #50 from RobDolinMS/patch-1

[ ReadMe ] Spelling fix

Commit:
Merge pull request #42 from wking/man-generate-options

man/ocitools-generate: Use [OPTIONS] in the synopsis

Commit:
Merge pull request #45 from wking/man-generate-template

man/ocitools-generate: Document --template

rhatdan committed:
Add mount label man page content

Signed-off-by: Dan Walsh <dwalsh@redhat.com>

Commit:
Merge pull request #52 from rhatdan/completions

Fixes for bash completions

Commit:
Merge pull request #53 from rhatdan/selinux

Add mount label man page content

liangchenye committed:
verify os and arch, following golang doc

Signed-off-by: liang chenye <liangchenye@huawei.com>

caniszczyk committed:
Update contributing guidelines and add maintainer guidelines

Each project in the OCI should have similar contributing guidelines
and maintainer guidelines.

Signed-off-by: Chris Aniszczyk <caniszczyk@gmail.com>

Commit:
Merge pull request #56 from opencontainers/update-contributing-guidelines

Update contributing guidelines and add maintainer guidelines

Commit:
Merge pull request #57 from liangchenye/osarch

verify os and arch, following golang doc

wking committed:
generate: Set ociVersion if it wasn't set in the template

The default template has this information already, but user-supplied
templates may not.  Before this commit:

  $ ocitools generate --template <(echo '{}')
  $ head -n2 config.json
  {
    "ociVersion": "",

And after this commit, the last line is:

    "ociVersion": "0.6.0-dev",

Regardless of this commit, template authors can set whatever version
they like without getting clobbered:

  $ ocitools generate --template <(echo '{"ociVersion": "xyz"}')
  $ head -n2 config.json
  {
    "ociVersion": "xyz",

Signed-off-by: W. Trevor King <wking@tremily.us>

wking committed:
generate: Set process.args to an empty array if it was nil

Before this commit:

  $ ocitools generate --template <(echo '{}')
  $ grep args config.json
      "args": null,

With this commit:

  $ grep args config.json
      "args": [],

That's still not a particularly useful value for the process
arguments, but it complies with the specs "array of strings" typing
[1] and is closer to a useful entry (a non-empty array).

[1]: https://github.com/opencontainers/runtime-spec/blob/dbce512cec0cd0774b236980ec18a4f6a91bc77e/config.md#process-configuration

Signed-off-by: W. Trevor King <wking@tremily.us>

wking committed:
generate: Don't add namespaces by default

And actively remove them if the PATH is the keyword "host".

This helps with --template, because it's harder to get clean template
round-trips if you have to opt-out of the namespace additions.  The
default template includes the non-user namespaces, so this commit
doesn't effect the bare 'ocitools generate' output.

Also document the PATH argument in ocitools-generate(1).  I couldn't
find a way to support '--network', etc. in codegangsta/cli, so the:

  --network=[PATH]

and such are intentional.  I would have preferred --network[=PATH],
but using --network= instead of --network for "create a new network
namespace" isn't that bad.

Signed-off-by: W. Trevor King <wking@tremily.us>

wking committed:
generate: Add replaceOrAppendNamespace to avoid clobbering template

If the template already has namespaces in it, we don't want to blow
them away in setupNamespaces.  For example:

  $ echo '{"linux": {"namespaces": [{"type": "pid"}]}}' >config-template.json
  $ ocitools generate --template config-template.json --network=abc
  $ jq .linux.namespaces config.json
  [
    {
      "type": "pid"
    },
    {
      "type": "network",
      "path": "abc"
    }
  ]
  $ ocitools generate --template config-template.json
  $ jq .linux.namespaces config.json
  [
    {
      "type": "pid"
    }
  ]

Signed-off-by: W. Trevor King <wking@tremily.us>

wking committed:
generate: Add --user and consolidate user-namespace handling

Put this in setupNamespaces with the other namespaces.  This commit
allows users to:

* Join an existing user namespace with --user=path/to/ns.
* Create a new user namespace without mapping IDs (although this is
  likely not very useful).
* Clear a templated user namespace with --user=host (although without
  being able to clear the ID mappings, this may not be very useful).

I haven't checked for likely-invalid configuration like:

  --uidmappings=1000:0:1 --user=path/to/ns

We can add that in a follow-up commit if we want.

Signed-off-by: W. Trevor King <wking@tremily.us>

rhatdan committed:
Fixup man pages for consistency

Some values on man pages do not show correctly, some indicate
multiple values when there is only a single value.  Some option
values were highlighted while others were not.

Signed-off-by: Dan Walsh <dwalsh@redhat.com>

Commit:
Merge pull request #58 from wking/default-version

generate: Set ociVersion if it wasn't set in the template

Commit:
Merge pull request #59 from wking/empty-array-default-for-process.args

generate: Set process.args to an empty array if it was nil

rhatdan committed:
--privileged should turn off defaults for SELinux, Apparmor and seccomp

Signed-off-by: Dan Walsh <dwalsh@redhat.com>

Commit:
Merge pull request #64 from rhatdan/privileged

--privileged should turn off defaults for SELinux, Apparmor and seccomp

Commit:
Merge pull request #65 from rhatdan/selinux

SELinux: Fixup man page

Commit:
Merge pull request #60 from wking/namespace-setup

generate: Assorted namespace cleanups (add --user, no clobber, ...)

liangchenye committed:
fix #67: verify if config is encoded in utf8

Signed-off-by: liang chenye <liangchenye@huawei.com>

liangchenye committed:
verify if cwd is an absolute path and env is valid

Signed-off-by: liang chenye <liangchenye@huawei.com>

wking committed:
man/ocitools-generate: Reword --poststop (after container-process exit)

"is stopped" echoes the spec [1].  I'd be a bit happier if the spec
used "exits" instead of "is stopped" (to avoid confusion with
SIGSTOP), but this man page is not the place to be generating new
language.

[1]: https://github.com/opencontainers/runtime-spec/blob/v0.5.0/config.md#poststop

Signed-off-by: W. Trevor King <wking@tremily.us>

Commit:
Merge pull request #71 from wking/post-stop-process-wording

man/ocitools-generate: Reword --poststop (after container-process exit)

Commit:
validation: add hostname validation

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

wking committed:
validate: Drop mount checks

These landed as CheckMounts in 647e3552 (bundle validate update to
0.3.0, 2016-02-23, #20), but both checks are too strict.

The first (destination exists in the rootfs) errors on valid cases
like:

  "mounts": [
    {
      "source": "users",
      "destination": "/home",
      "type": "bind"
    },
    {
      "source": "none",
      "destination": "/home/wking",
      "type": "tmpfs"
    }
  ]

Where the source 'users' directory already contained a 'wking'
subdirectory.  So by the time the tmpfs was setup, the destination
directory would exist, but at validation time (without having run the
bind mount) the tmpfs destination directory would not exist.

The second (destination is a directory) errors on valid cases like:

  "mounts": [
    {
      "source": "/etc/resolv.conf",
      "destination": "/etc/resolv.conf",
      "type": "bind"
    }
  ]

because binding files to files works.  In a shell:

  # touch test
  # mount --bind /etc/resolv.conf test
  # umount test

However binding directories to files does not work:

  # mount --bind /etc test
  mount: mount point /tmp/test is not a directory

Figuring out which mount configurations are valid and which aren't may
be possible, but I'm pretty sure it's more trouble than we want to get
into.  There may be room for other mount tests (e.g. comparing 'type'
against /proc/filesystems as a host-specific test), but I'm leaving
those to subsequent pull requests.

Signed-off-by: W. Trevor King <wking@tremily.us>

Commit:
Merge pull request #70 from liangchenye/utf8

verify: config is encoded in utf8; cwd is absolute path; env is valid

Commit:
Merge pull request #73 from wking/drop-mount-checks

validate: Drop mount checks

Commit:
Merge pull request #72 from Mashimiao/add-hostname-bundle-validation

validation: add hostname validation

Commit:
checkLinux executed when platform set as linux

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

Commit:
Merge pull request #49 from Mashimiao/add-hooks-bundle-validation

validate: add Hooks validation

Commit:
runtimetest: add readonly path validation

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

Commit:
ocitools/man: small fix

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

Commit:
Merge pull request #77 from Mashimiao/man-tfix-fortmat-fix

ocitools/man: small fix

Commit:
runtimetest: add root filesystem validation

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

Commit:
runtimetest/hostname: small fix for null hostname

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

Commit:
Merge pull request #79 from Mashimiao/bugfix-runtimetest-hostname

runtimetest/hostname: small fix for null hostname

Commit:
runtimetest: add masked paths validation

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

Commit:
Merge pull request #78 from Mashimiao/runtime-test-root-fs-validation

runtimetest: add root filesystem validation

Commit:
validation: add env validation for hooks

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

Commit:
add --hooks option for validate completion

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

Commit:
validate: small fix for root path error

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

Commit:
Merge pull request #84 from Mashimiao/add-env-valid-for-hooks

validation: add env validation for hooks

Commit:
Merge pull request #86 from Mashimiao/fix-validate-root-path-error

validate: small fix for root path error

Commit:
Merge pull request #82 from Mashimiao/runtime-test-maksed-path-validation

runtimetest: add masked paths validation

Commit:
Merge pull request #85 from Mashimiao/completions-add-hooks-for-validate

add --hooks option for validate completion

Commit:
Merge pull request #75 from Mashimiao/sepcified-when-to-execute-checklinux

just check hostname with UTS namespace when platform set as linux

Commit:
Merge pull request #76 from Mashimiao/runtime-test-readonly-path-validation

runtimetest: add readonly path validation

Commit:
ocitols mans: tfix

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

Commit:
runtimetest: optimize capabilities validation

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

Commit:
Merge pull request #87 from Mashimiao/man-tfix-and-uppercase-fix

ocitols mans: tfix

Commit:
Fix typo in man/ocitools-generate.1.md

Signed-off-by: Haiyan Meng <hmeng@redhat.com>

Commit:
Merge pull request #90 from hmeng-19/fix_man_typo

Fix typo in man/ocitools-generate.1.md

Commit:
generate: add specified cgroups path support

Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

Commit:
Merge pull request #94 from Mashimiao/generate-add-cgroups-path-support

generate: add specified cgroups path support

Commit:
Merge pull request #89 from Mashimiao/runtime-test-optimize-cap-validation

runtimetest: optimize capabilities validation

liangchenye committed:
output all the detected errors

Signed-off-by: liang chenye <liangchenye@huawei.com>
